### ISSUE TITLE
Support send-only endpoints

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus.TransactionalSession#96]
The transaction session can now be used in send-only endpoints. See our [documentation for more details](https://docs.particular.net/nservicebus/transactional-session/#send-only).

### Context

Currently, the transactional session requires the endpoint using the transactional session to be an endpoint with a queue. Send-only endpoints are not supported. One of the restrictions is given by the Core since the [outbox seam prevents it being used in send-only endpoints](https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Reliability/Outbox/Outbox.cs#L20-L21). 

By lifting these constraints (potentially with a settings bypass) we could separate the endpoint configuration into two parts:

- Enabling the sending part, including a logical address of the "owning endpoint" (potentially the backend for the frontend). This would also instruct Core to lift the outbox constraint mentioned above
- Enabling the receiving part (the control message behaviors) on the "owning endpoint" (potentially the backend for the frontend)

**Comment by pardahlman** [comment]
I'm looking forward using this package and get rid of complicated code written to get consistency between NServiceBus message publishing and database updates in our ASP.NET Core applications.

However, it surprises me that it is not possible to use this package in a send-only endpoint yet. When reading the documentation and [blog posts](https://particular.net/blog/transactional-session) it sounds like just the scenario you're aiming for.

Are there any estimates on when send-only endpoints will be supported?

**Comment by SzymonPobiega** [comment]
@pardahlman no estimates currently. To make sure we understand you use case better -- is there anything that would prevent you from using full NServiceBus endpoint in the web application?

**Comment by pardahlman** [comment]
Thanks for getting back to me 🙏

No, there is no technical limitation, rather a logical one: these endpoint does not consume messages and is therefor correctly configured as send-only endpoints. Making them full endpoints would slow down startup slightly and also create unwanted topology in our transport.

**Comment by SzymonPobiega** [comment]
OK, I understand.

The current limitation is related to the fact, that the transactional session uses a message send locally to commit or abort the transaction. This special control message plays the role which is fulfilled by a database polling background process in other outbox implementation. We decided to use the message because the polling mechanism causes serval issue in any non-trivial production deployment, especially with partition data stores:
 - polling a database when there is no traffic is a waste of resource
 - polling might not be easy when using serverless hosting (or requires a dedicated timer-based process)
 - polling partition data stores with a query that needs to read all partitions is _very_ costly

Thanks for your feedback, we are definitely going to take it into account when releasing new versions. We will again analyze the consequences of introducing (optional) polling or having a dedicated endpoint processing the control messages so that the web frontend endpoints can stay send-only.

**Comment by danielmarbach** [comment]
Please allow me to add a bit more context here too. We have also discussed that we could allow sending the control message to a dedicated "backend" endpoint. During the initial design discussions, we have discussed the concept of ownership of endpoint data and how that "backend" endpoint would fit it. We brought up the argument that many times API endpoints that are send-only have some sort of a backend endpoint already and that one has to be treated as part of the "same service". So, ownership of the outbox data should not be too much of a problem. Yet because we didn't entirely settle on a viable approach on these discussions, we have decided to "scope it out" from the initial release. I'm mentioning this because we don't necessarily need to introduce polling. Sending the control message to somewhere else is totally feasible too.

**Comment by hugovlc** [comment]
Now that the backdoor setting that allows this to be implemented is done, do you guys have a planned milestone to implement the feature in the near future?

**Comment by andreasohlund** [comment]
Hi @hugovlc this will likely be picked up by our recurring enhancement release process but I can't give you a timeline at this stage. 

How does the timeline look on your end? Are you blocked by this not being available?

**Comment by hugovlc** [comment]
Yes because we recently changed transport to SqlServer transport and our business data is in oracle. TransactionScope is not an option.
So there is no way now to send messages and save business data atomicaly. 
We cannot change the endpoint to a full endpoint because then is changed as an extra one.

**Comment by alexander-kucherov** [comment]
It really cool feature. I'm looking forward for it implementation.
Could you please provide an update on the progress of implementing the functionality?

**Comment by mikeminutillo** [comment]
> It really cool feature. I'm looking forward for it implementation. Could you please provide an update on the progress of implementing the functionality?

There's nothing new to report since [this last update](https://github.com/Particular/NServiceBus.TransactionalSession/issues/96#issuecomment-1657737504). We are tracking this in our backlog, but we have not started the implementation yet.

**Comment by patrick-l-boe** [comment]
FWIW my team is waiting on this feature before we can stop relying on TransactionScope. This is holding us back from support on some platforms.

**Comment by andreasohlund** [comment]
Hi all, I wanted to update you on where we stand on this.

1. We have decided to schedule technical work to make send-only endpoints to be able to use the transactional session. I have no firm timeline yet but we will let you know once we start to make progress
2. In the interim we will make sure that you are not paying for the endpoints that you have to (temporarily) convert to full endpoints as a workaround. Please contact us and we'll ensure those endpoints get excluded from license costs.

**Comment by daenumueller** [comment]
Hi, any update on this topic and a possible timeline?

**Comment by ThreePinkApples** [comment]
Is there any status update on this? It's been a while since the end of Q1

**Comment by andreasohlund** [comment]
@ThreePinkApples sorry for the delay, we had to adjust our priorities but the work  is now set to begin within 1-2 months so my best guestimate as of now is late Q4 this year

Are you in any way blocked by this?

**Comment by ThreePinkApples** [comment]
@andreasohlund No, not really blocked by this, more a nice to have for us in some cases. Just wanted to know whether it had been abandoned or not

**Comment by andreasohlund** [comment]
>just wanted to know whether it had been abandoned or not

Got it, thanks for taking the time to reach out. We'll keep you posted here as we start working!

**Comment by jbogard** [comment]
Hit this today and was at first surprised it wasn't a supported scenario. Then remembered how hard it is to implement a robust outbox and I've done entire talks on it. Don't worry, I'll still complain it's not there!

**Comment by andreasohlund** [comment]
Thanks for the reminder @jbogard :)

The current plan is to start this work in the next couple of months so stay tuned

**Comment by sticl** [comment]
This would be very useful indeed, we need various specialized message sessions and implementations just to support the outbox functionality:

In a receiving endpoint:
- Inside a handler: regular message session 
- Outside of the context of a handler: Use of transactional session (this package).

In a send-only endpoint:
- We rely on our custom-built outbox.

Adding to the complexity, the outbox behavior differs across implementations:

- Regular UniformSession: Immediately flushes the outbox as part of the message handling.
- TransactionalSession: Uses a control message to send to the same queue after which it flushes the outbox (on crowdy queues can be totally different behavior in timing for flushing the outbox).
- Custom-built outbox: relies on background jobs, as we need to store messages in a separate table.

Uniforming outbox behavior in send-only endpoints & in receiving endpoints outside of a handler would help a lot.

**Comment by timbussmann** [comment]
bumping this issue as well.

A special use case where using send-only endpoints would be be helpful are scenarios with short-lived publishers that immediately shutdown after committing the tx session. There is no safety guard that ensures the control message will actually be consumed, which could even lead to a full message loss in extreme situations.
Using a queue of another endpoint that shares the same outbox would be a simple approach to ensure that control message will be consumed within reasonable time. To do that the publisher needs to be send-only only to avoid consuming from the same queue though, which is currently not supported.

**Comment by andreasohlund** [comment]
@timbussmann agreed, this use case in included in the scope of the work about to start, we'll keep you posted here as we make progress

**Comment by andreasohlund** [comment]
We've started work on this and will report back here as we progress

**Comment by andreasohlund** [comment]
We are making some progress on this feature, the API we're looking at is 

```
        var transactionalSessionOptions = new TransactionalSessionOptions
        {
            ProcessorAddress = "MyRemoteProcessor"
        };

        endpointConfiguration.UsePersistence<MyPersistence>()
            .EnableTransactionalSession(transactionalSessionOptions);
```

to enable send-only endpoints to configure a remote processing endpoint that will handle the control message used to drive the dispatch of the outgoing messages.

Work in progress docs - https://github.com/Particular/docs.particular.net/pull/7089/

**Comment by jbogard** [comment]
@andreasohlund what would be the requirements of this control host? Like would it need references to the messages or originating app etc.

**Comment by andreasohlund** [comment]
>what would be the requirements of this control host? Like would it need references to the messages or originating app etc.

Only thing needed is to have the outbox enabled and connected to the same database. Does that make sense?

**Comment by jbogard** [comment]
Yep that does, thanks!

**Comment by andreasohlund** [comment]
This is done and will be released shortly. See https://docs.particular.net/nservicebus/transactional-session/#send-only for instructions on how to use it

**Comment by DorianGreen** [comment]
@andreasohlund are there any plans to back-port this functionality to v2 so it can be used by NSB v8 on projects stuck on .NET framework?

If not, will those endpoints remain exempt from licensing if they are only used for the transactional outbox?
